Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-57.ca-central-1.compute.internal
Sha: 0145ac065830773a06d4d080d619dcf3a8f211cb
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-13627/HPCC-Platform/rtl/eclrtl/rtlrecord.cpp:153:111: error: enumeral and non-enumeral type in conditional expression [-Werror=extra]
cc1plus: all warnings being treated as errors
make[2]: *** [rtl/eclrtl/CMakeFiles/eclrtl.dir/build.make:271: rtl/eclrtl/CMakeFiles/eclrtl.dir/rtlrecord.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:16200: rtl/eclrtl/CMakeFiles/eclrtl.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/cc57cf2d5ece51a09019e92fe512342a)
Build log [Build log](https://gist.github.com/HPCCSmoketest/cc57cf2d5ece51a09019e92fe512342a#file-result-20-04-14-14-55-07-log)

--- a/rtl/eclrtl/rtlrecord.cpp
+++ b/rtl/eclrtl/rtlrecord.cpp
@@ -149,8 +149,8 @@ private:
 class RtlCondFieldStrInfo : public RtlFieldStrInfo
 {
 public:
-    RtlCondFieldStrInfo(const RtlFieldInfo &from, const IfBlockInfo &_ifblock)
-    : RtlFieldStrInfo(from.name, from.xpath, from.type, from.flags | (RFTMinifblock|RFTMdynamic), (const char *) from.initializer),
+    RtlCondFieldStrInfo(const RtlFieldInfo &from, const IfBlockInfo &_ifblock, bool forcePayload)
+    : RtlFieldStrInfo(from.name, from.xpath, from.type, from.flags | (forcePayload ? RFTMinifblock|RFTMdynamic|RFTMispayloadfield : RFTMinifblock|RFTMdynamic), (const char *) from.initializer),
       origField(from),ifblock(_ifblock)
     {
     }
@@ -159,7 +159,7 @@ public:
     const IfBlockInfo &ifblock;
 };
 
-static unsigned expandNestedRows(unsigned idx, unsigned startIdx, StringBuffer &prefix, StringBuffer &xPathPrefix, const RtlFieldInfo * const * fields, const RtlFieldInfo * * target, const char * *names, const char * *xpaths, const IfBlockInfo *inIfBlock, ConstPointerArrayOf<IfBlockInfo> &ifblocks)
+static unsigned expandNestedRows(unsigned idx, unsigned startIdx, StringBuffer &prefix, StringBuffer &xPathPrefix, const RtlFieldInfo * const * fields, const RtlFieldInfo * * target, const char * *names, const char * *xpaths, const IfBlockInfo *inIfBlock, ConstPointerArrayOf<IfBlockInfo> &ifblocks, bool forcePayload)
 {
     for (;*fields;fields++)
     {
@@ -188,7 +188,7 @@ static unsigned expandNestedRows(unsigned idx, unsigned startIdx, StringBuffer &
                     if (!isEmptyString(xpath))
                         xPathPrefix.append(xpath).append('/');
                 }
-                idx = expandNestedRows(idx, isIfBlock ? startIdx : idx, prefix, xPathPrefix, nested, target, names, xpaths, nestIfBlock, ifblocks);
+                idx = expandNestedRows(idx, isIfBlock ? startIdx : idx, prefix, xPathPrefix, nested, target, names, xpaths, nestIfBlock, ifblocks, forcePayload || (cur->flags & RFTMispayloadfield) != 0);
                 prefix.setLength(prevPrefixLength);
                 if (xpaths)
                     xPathPrefix.setLength(prevXPathPrefixLength);
@@ -215,7 +215,12 @@ static unsigned expandNestedRows(unsigned idx, unsigned startIdx, StringBuffer &
                     xpaths[idx] = nullptr;
             }
             if (inIfBlock && !(cur->flags & RFTMinifblock))
-                target[idx++] = new RtlCondFieldStrInfo(*cur, *inIfBlock);
+                target[idx++] = new RtlCondFieldStrInfo(*cur, *inIfBlock, forcePayload);
+            else if (forcePayload && !(cur->flags & RFTMispayloadfield))
+            {
+                dbgassertex((cur->flags & RFTMdynamic) == 0);
+                target[idx++] = new RtlFieldStrInfo(cur->name, cur->xpath, cur->type, cur->flags | (RFTMispayloadfield|RFTMdynamic), (const char * ) cur->initializer);
+            }
             else
             {
                 dbgassertex((cur->flags & RFTMdynamic) == 0);
@@ -272,7 +277,7 @@ RtlRecord::RtlRecord(const RtlFieldInfo * const *_fields, bool expandFields) : f
                 xpaths = new const char *[numFields];
             fields = allocated;
             StringBuffer prefix, xPathPrefix;
-            unsigned idx = expandNestedRows(0, 0, prefix, xPathPrefix, originalFields, allocated, names, xpaths, nullptr, _ifblocks);
+            unsigned idx = expandNestedRows(0, 0, prefix, xPathPrefix, originalFields, allocated, names, xpaths, nullptr, _ifblocks, false);
             ifblocks = _ifblocks.detach();
             assertex(idx == numFields);
             allocated[idx] = nullptr;
@@ -585,7 +590,7 @@ const RtlRecord *RtlRecord::queryNested(unsigned fieldId) const
 const RtlFieldInfo *RtlRecord::queryOriginalField(unsigned idx) const
 {
     const RtlFieldInfo *field = queryField(idx);
-    if (field->flags & RFTMdynamic)
+    if (field->flags & RFTMinifblock)
         return &static_cast<const RtlCondFieldStrInfo *>(field)->origField;
     else
         return field;

